### PR TITLE
Add `DfTx` AppointOracle/RemoveOracle/UpdateOracle/SetOracleData

### LINKED ISSUE #201: `DeFiCh/ain:1.7.x` DeFi CustomTx (DfTx)
Similar to https://github.com/DeFiCh/jellyfish/issues/108

Implementation of 1.7.x Custom DfTx
https://github.com/DeFiCh/ain/blob/6428ac0ca470dd16f973d62b74a1ec0b9e5444a8/src/masternodes/mn_checks.h#L62-L67

```cpp
enum class CustomTxType : uint8_t
{
  AppointOracle         = 'o',
  RemoveOracleAppoint   = 'h',
  UpdateOracleAppoint   = 't',
  SetOracleData         = 'y',
}
```

in

```
packages/jellyfish-transaction/src/script/defi/*
```

All CustomTX has a starting signature of `DfTx` in utf8 or `0x44665478`. It is as it is, don't need to convert from BE to LE.

They are structured as a 
1. `VarUInt` + [`OP_RETURN` + `DfTx` + `TYPES[CxTxType]` + ...]
1. `VarUInt` + `6a` + `44665478` + Type + ...

Refer to #108 for a detailed implementation guide

### PR DESCRIPTION
**Pull request by monstrobishi**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Implementation of 1.7.x Custom DfTx
https://github.com/DeFiCh/ain/blob/6428ac0ca470dd16f973d62b74a1ec0b9e5444a8/src/masternodes/mn_checks.h#L62-L67

```
enum class CustomTxType : uint8_t
{
  AppointOracle         = 'o',
  RemoveOracleAppoint   = 'h',
  UpdateOracleAppoint   = 't',
  SetOracleData         = 'y',
}
```

#### Which issue(s) does this PR fixes?:

Fixes #201

### CODE REVIEW

**Comment by monstrobishi** [comment]
This is a draft PR, so far `AppointOracle` has been addded

**Review by canonbrother — APPROVED** [review]

File: packages/testcontainers/src/chains/defid_container.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ export interface StartOptions {
 export abstract class DeFiDContainer extends DockerContainer {
   /* eslint-disable @typescript-eslint/no-non-null-assertion, no-void */
   public static readonly PREFIX = 'defichain-testcontainers-'
-  public static readonly image = 'defi/defichain:1.6.4'
+  public static readonly image = 'defi/defichain:HEAD-75b1420'
```
Comment (fuxingloh) [inline packages/testcontainers/src/chains/defid_container.ts:25]:
```suggestion
  public static readonly image = 'defi/defichain:1.6.4'
```

revert this

**Review by fuxingloh — CHANGES_REQUESTED** [review]
haven't checked everything but 1.6.4 should not be changed yet.

File: packages/testcontainers/src/chains/defid_container.ts (hunk 1 above)
Comment (monstrobishi) [reply packages/testcontainers/src/chains/defid_container.ts:25]: thanks for spotting! that's true this PR doesn't actually need the new node version

File: packages/jellyfish-transaction/src/script/defi/dftx_price.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import BigNumber from 'bignumber.js'
```
Comment (ivan-zynesis) [inline packages/jellyfish-transaction/src/script/defi/dftx_price.ts:1]: Perhaps should test `dftx_price` composable individually, the interfaces looks very generic and applicable for different usage.

Comment (monstrobishi) [reply packages/jellyfish-transaction/src/script/defi/dftx_price.ts:1]: it's only used for OracleTx, it's a very specific structure

File: packages/jellyfish-transaction/src/script/defi/dftx_price.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    return [
+      ComposableBuffer.varUIntUtf8BE(() => cp.token, v => cp.token = v),
+      ComposableBuffer.varUIntUtf8BE(() => cp.currency, v => cp.currency = v)
+    ]
+  }
+}
+
+export interface TokenAmount {
+  currency: string // ---------------------| max 8 bytes
+  amount: BigNumber // -----------------| 8 bytes unsigned
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_price.ts:27]: 1. fix the indentation align the `---|` properly.
2. Be extra specific, too much guessing work here, I have no idea how to parse it just looking at these comments 
3. Look at other BufferComposer examples

File: packages/jellyfish-transaction/src/script/defi/dftx_price.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+   * Composable TokenAmount, C stands for Composable.
+   * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
+   */
+export class CTokenAmount extends ComposableBuffer<TokenAmount> {
+  composers (tp: TokenAmount): BufferComposer[] {
+    return [
+      ComposableBuffer.varUIntUtf8BE(() => tp.currency, v => tp.currency = v),
+      ComposableBuffer.satoshiAsBigNumber(() => tp.amount, v => tp.amount = v)
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_price.ts:38]: are you sure it's a satoshi unit? satoshiAsBigNumber is only used for 16 bytes, `0.00000000` calculation.

File: packages/jellyfish-transaction/src/script/defi/dftx_oracles.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// Disabling no-return-assign makes the code cleaner with the setter and getter */
+/* eslint-disable no-return-assign */
+
+/**
+ * AppointOracle DeFi Transaction
+ */
+export interface AppointOracle {
+  script: Script // --------------------| n = VarUInt{1-9 bytes}, + n bytes
+  weightage: number // -------------------| 1 byte unsigned
+  pricefeeds: CurrencyPair[] // -----------------| array of strings each max 8 bytes
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_oracles.ts:16]: use camelCase

**Review by fuxingloh — CHANGES_REQUESTED** [review]
Your oracle ComposableBuffer setup needs to be extremely explicit. I must be able to look at the interface and decode it without looking at `composers`.

And fix the indentations.

https://github.com/DeFiCh/jellyfish/blob/63fd7700ef9959cc50bb2dadfc3cff202771aaab/packages/jellyfish-transaction/src/tx.ts#L60-L66

I haven't reviewed everything, there was too many inconsistency.

File: packages/jellyfish-transaction/src/script/defi/dftx_price.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+export interface TokenAmount {
+  currency: string // ------------------| c = VarUInt{1-9 bytes}, + c bytes UTF encoded string
+  amount: BigNumber // -----------------| 8 bytes
+}
+
+/**
+   * Composable TokenAmount, C stands for Composable.
+   * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
+   */
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_price.ts:33]:
```suggestion
/**
 * Composable TokenAmount, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
```

File: packages/jellyfish-transaction/src/script/defi/dftx_price.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+export interface TokenPrice {
+  token: string // ---------------------| c = VarUInt{1-9 bytes}, + c bytes UTF encoded string
+  prices: TokenAmount[] // -------------| c = VarUInt{1-9 bytes}, + c x TokenAmount
+}
+
+/**
+   * Composable TokenPrice, C stands for Composable.
+   * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
+   */
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_price.ts:51]:
```suggestion
/**
 * Composable TokenPrice, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]
minor JSDocs changes.

---

If I am not wrong, I remember you documented about `token` & `currency` max length to be 8. If that's the case, we should probably assert this behaviour at the buffer composer level.

Without assertions, this is a potentially **very dangerous situation due to Buffer overflow.** 

All BufferComposer methods are actually asserted for overflow and tested. Examples are below for reference.

https://github.com/DeFiCh/jellyfish/blob/241f3906ea43d1f32974d0c3dd7fded75627582b/packages/jellyfish-transaction/__tests__/buffer/buffer_composer.test.ts#L646-L652

https://github.com/DeFiCh/jellyfish/blob/241f3906ea43d1f32974d0c3dd7fded75627582b/packages/jellyfish-transaction/__tests__/buffer/buffer_composer.test.ts#L234-L236

**Comment by monstrobishi** [comment]
@fuxingloh 

Will remove comment regarding string length instead, the string is not stored in a fixed size buffer, and it gets trimmed at both RPC and consensus level at the node, but this is not expected as part of the tx composition, there is no risk of buffer overflow.

**Comment by monstrobishi** [comment]
Looks like the string length comment was removed already. I can check this in txn-builder to mimic RPC behaviour, but there is no restriction on buffer level.

**Review by fuxingloh — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

**Pull request merged**